### ISSUE TITLE
Testing comboBox selections

### ISSUE THREAD

**Issue by aalmiray** [TestFX/TestFX#216]
I've looked at the examples and the main project for clues to perform the following scenario:

    given: "a combobox with some items"
    ComboBox cmb1 = new ComboBox()
    cmb1.items = ([0..5].collect { 'item ' + it}) as ObservableList
    cmb1.id = 'cmb1'

    when: "the combobox has the focus"
    clickOn('#cmb1')

    and: "the third entry is selected"
    ???

    then:
    cmb1.selectionModel.selectedItem == 'item 2'

Basically, how would you scroll down the list of items and perform a selection using the robot?

**Comment by hastebrot** [comment]
Hi andres,

is your combobox item out of sight? This would mean we need to find a
method to scroll its internal scroll pane.

**Comment by hastebrot** [comment]
If the item is visible I simply would call onClick("item 2").

**Comment by aalmiray** [comment]
Benjamin: the comboBox is visible. Invoking "clickOn('item')" would work as long as there is no other clickable node that has the same text, iow, addign a button with text "item 2" would make this test fail, wouldn't it?

AFAIK clickOn(text) would try to match a node from the root, but in this case I'd liek to match a visible element given a narrower context: the comboBox. Is this possible?

Additionally, the ability to scroll the internal SP would be great as I can see this kind of behavior also needed in Lists and other scrollable nodes. I think having the ability to click on an item given an (scrollable) index would be agood thing to have, besides having a text matcher as it currently stands today.

WDYT?

**Comment by jairocgr** [comment]
I'm having the same problem with ChoiceBoxes :/

**Comment by hastebrot** [comment]
I've experimented a bit with a function that scrolls to nodes outside of the visible area.

One problem (for `ListView`, `TableView`, `TreeView`, and `TreeTableView`) is, that you can't query them by text, since their text is lazily created.

**Comment by stepasite** [comment]
I face the similar issue. I cannot select combox item by its label as another GUI element containing the same text exists in the GUI. Any help how to select such item (by mouse cursor) would be much appreciated.

**Comment by arnobl** [comment]
May help: I implemented routines to interact with JFX combo boxes:
https://github.com/arnobl/latexdraw/blob/master/latexdraw-core/net.sf.latexdraw/src/test/java/net/sf/latexdraw/instruments/robot/FxRobotListSelection.java

**Comment by brcolow** [comment]
Making it easier to select single and multiple items (via JavaFX SelectionModels) is something I definitely want to work on in TestFX. Currently I am struggling with how best to fit this into the current API. Adding `select` methods to `FxRobot` seems like a good idea.

**Comment by grigala** [comment]
I'm currently also trying to test `ComboBox` selections. Tried a few alternatives. So far the only thing that worked for me was `type(xTimes, KeyCode.DOWN)`.

First tried to use `comboBox.getSelectionModel().select(INDEX)` which seems to be working in terms of selection but still throws an 
<details>
  <summary>IllegalStateException</summary>

  ```java
 --- Exception in Async Thread ---
java.lang.IllegalStateException: Not on FX application thread; currentThread = Test worker
	javafx.graphics/com.sun.javafx.tk.Toolkit.checkFxUserThread(Toolkit.java:292)
	javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.checkFxUserThread(QuantumToolkit.java:446)
	javafx.graphics/javafx.scene.Parent$3.onProposedChange(Parent.java:474)
	javafx.base/com.sun.javafx.collections.VetoableListDecorator.setAll(VetoableListDecorator.java:113)
	javafx.base/com.sun.javafx.collections.VetoableListDecorator.setAll(VetoableListDecorator.java:108)
	javafx.controls/javafx.scene.control.skin.LabeledSkinBase.updateChildren(LabeledSkinBase.java:272)
	javafx.controls/javafx.scene.control.skin.LabeledSkinBase.lambda$new$11(LabeledSkinBase.java:220)
	javafx.controls/com.sun.javafx.scene.control.LambdaMultiplePropertyChangeListenerHandler.lambda$new$1(LambdaMultiplePropertyChangeListenerHandler.java:49)
	javafx.base/javafx.beans.value.WeakChangeListener.changed(WeakChangeListener.java:86)
	javafx.base/com.sun.javafx.binding.ExpressionHelper$SingleChange.fireValueChangedEvent(ExpressionHelper.java:181)
	javafx.base/com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:80)
	javafx.base/javafx.beans.property.StringPropertyBase.fireValueChangedEvent(StringPropertyBase.java:104)
	javafx.base/javafx.beans.property.StringPropertyBase.markInvalid(StringPropertyBase.java:111)
	javafx.base/javafx.beans.property.StringPropertyBase.set(StringPropertyBase.java:145)
	javafx.base/javafx.beans.property.StringPropertyBase.set(StringPropertyBase.java:50)
	javafx.base/javafx.beans.property.StringProperty.setValue(StringProperty.java:65)
	javafx.controls/javafx.scene.control.Labeled.setText(Labeled.java:147)
	javafx.controls/javafx.scene.control.skin.ComboBoxListViewSkin.updateDisplayText(ComboBoxListViewSkin.java:459)
	javafx.controls/javafx.scene.control.skin.ComboBoxListViewSkin$2$1.updateItem(ComboBoxListViewSkin.java:495)
	javafx.controls/javafx.scene.control.ListCell.updateItem(ListCell.java:478)
	javafx.controls/javafx.scene.control.ListCell.indexChanged(ListCell.java:337)
	javafx.controls/javafx.scene.control.IndexedCell.updateIndex(IndexedCell.java:120)
	javafx.controls/javafx.scene.control.skin.ComboBoxListViewSkin.updateDisplayNode(ComboBoxListViewSkin.java:352)
	javafx.controls/javafx.scene.control.skin.ComboBoxListViewSkin.getDisplayNode(ComboBoxListViewSkin.java:266)
	javafx.controls/javafx.scene.control.skin.ComboBoxBaseSkin.updateDisplayArea(ComboBoxBaseSkin.java:289)
	javafx.controls/javafx.scene.control.skin.ComboBoxBaseSkin.lambda$new$7(ComboBoxBaseSkin.java:133)
	java.base/java.util.function.Consumer.lambda$andThen$0(Consumer.java:65)
  ```  
</details>

**Comment by hastebrot** [comment]
>still throws an `IllegalStateException`

@grigala Interactions with JavaFX components should be made within a lambda function using FxRobot#interact().

```java
interact(() -> {
  comboBox.getSelectionModel().select(INDEX);
});
```

Also: Greetings to former colleagues at Karakun.

**Comment by grigala** [comment]
> @grigala Interactions with JavaFX components should be made within a lambda function using FxRobot#interact().
> 
> ```java
> interact(() -> {
>   comboBox.getSelectionModel().select(INDEX);
> });
> ```

Thanks for the hint. 

> 
> Also: Greetings to former colleagues at Karakun.

Oh, will definitely do, cheers!
